src/python.cpp

```diff
@@ -1126,19 +1148,38 @@ SEXP py_to_r(PyObject* x, bool convert) {
 
       case NPY_STRING:
       case NPY_UNICODE: {
-        PyObjectPtr itemFunc(PyObject_GetAttrString(ptrArray, "item"));
-        if (itemFunc.is_null())
+
+        PyObjectPtr nditerArgs(PyTuple_New(1));
+        // PyTuple_SetItem steals reference the array, but it's already wraped
+        // into PyObjectPtr earlier (so it gets deleted after the scope of this function)
+        // To avoid trying to delete it twice, we need to increase its ref count here.
+        PyTuple_SetItem(nditerArgs, 0, (PyObject*)array);
+        Py_IncRef((PyObject*)array);
+
+        PyObjectPtr iter(PyObject_Call(get_np_nditer(), nditerArgs, NULL));
+
+        if (iter.is_null()) {
           throw PythonException(py_fetch_error());
+        }
+
         rArray = Rf_allocArray(STRSXP, dimsVector);
         RObject protectArray(rArray);
+
+        // empty tuple, casts to scalar value
+        PyObjectPtr itemArg(PyTuple_New(0));
         for (int i=0; i<len; i++) {
-          PyObjectPtr pyArgs(PyTuple_New(1));
-          // PyTuple_SetItem steals reference to object created by PyInt_FromLong
-          PyTuple_SetItem(pyArgs, 0, PyInt_FromLong(i));
-          PyObjectPtr pyStr(PyObject_Call(itemFunc, pyArgs, NULL));
+          PyObjectPtr el(PyIter_Next(iter)); // returns an scalar array.
+
+          PyObjectPtr itemFunc(PyObject_GetAttrString(el, "item"));
```

Comment:
Would it make sense to resolve the `item` method once from `np.ndarray.item`, before the loop, then call `item(el)` inside the loop?

Reply:
I figured I could use `PyObject_CallMethod` instead to avoid having to manually extract the class method.